frontend/controller/service-worker.js

```diff
@@ -13,23 +15,171 @@ sbp('sbp/selectors/register', {
     // TODO: move ahead with encryption stuff ignoring this service worker stuff for now
     // TODO: improve updating the sw: https://stackoverflow.com/a/49748437
     // NOTE: user should still be able to use app even if all the SW stuff fails
+    if (!('serviceWorker' in navigator)) { return }
+
     try {
-      await navigator.serviceWorker.register('/assets/js/sw-primary.js', {
-        scope: '/'
-      })
-      navigator.serviceWorker.addEventListener('message', function (event) {
-        console.info('Received message from SW:', event.data)
+      const swRegistration = await navigator.serviceWorker.register('/assets/js/sw-primary.js', { scope: '/' })
+
+      if (swRegistration) {
+        swRegistration.active?.postMessage({ type: 'store-client-id' })
+      }
+
+      navigator.serviceWorker.addEventListener('message', event => {
+        console.debug('[sw] Received a message from the service worker :', event)
+        const data = event.data
+
+        if (typeof data === 'object' && data.type) {
+          switch (data.type) {
+            case 'pushsubscriptionchange': {
+              sbp('service-worker/resubscribe-push', data.subscription)
+              break
+            }
+            default:
+              console.error('[sw] Received unknown message type from the service worker:', data)
+              break
+          }
+        }
       })
-      setTimeout(function () {
-        navigator.serviceWorker.controller.postMessage({
-          type: 'set',
-          key: 'foo',
-          value: 'bar'
-        })
-      }, 1000)
     } catch (e) {
-      // TODO: this
       console.error('error setting up service worker:', e)
     }
+  },
+  'service-worker/setup-push-subscription': async function (followingNotification = null) {
+    if (!('serviceWorker' in navigator) || !('Notification' in window)) { return }
+
+    // Get the installed service-worker registration
+    const registration = await navigator.serviceWorker.ready
+
+    if (!registration) {
+      console.error('No service-worker registration found!')
+      return
+    }
+
+    const pubsub = sbp('okTurtles.data/get', PUBSUB_INSTANCE)
+    const existingSubscription = await registration.pushManager.getSubscription()
+
+    if (existingSubscription) {
+      // If there is an existing subscription, no need to create a new one.
+      // But make sure server knows the subscription details too.
+      pubsub.socket.send(createMessage(
```

Comment:
Note that low-level `.socket.send()` calls do not benefit from automatic retriability upon reconnection. More specifically, the message won't be sent if `.socket.readyState !== WebSocket.OPEN`.
Not sure whether that's an actual problem though, I don't know enough about service workers.

Reply:
@snowteamer 
Currently where this sbp selector(`service-worker/setup-push-subscription`) is called is righter after the user's signup/login so at least, we can guarantee this won't be executed when the `socket.readyState === WebSocket. CONNECTING`.

But,

>  benefit from automatic retriability upon reconnection

sounds like an important stuff to have. If you give some explanations on what the "non" low-level usage of the pubsub system. I can make the modification in the follow-up PR while working on [this issue](https://github.com/okTurtles/group-income/issues/1791).

Reply:
@snowteamer A reminder for you for a potentially important question I left to you above, if this has gone unnoticed.

Reply:
@SebinSong 
>  So at least, we can guarantee this won't be executed when the socket.readyState is still === WebSocket.CONNECTING
Sounds fine to me - maybe we need to check for `WebSocket.CONNECTING`?
Also do we need to call that selector again upon reconnection after a temporary disconnection?

> what the "non" low-level usage of the pubsub system would be
That would be using client methods like `sub()` or `pub()` instead of socket methods, but not sure whether there is a currently fitting one for this use case

Reply:
@snowteamer Thanks for your opinion. I think the entire service-worker related workflow added by this PR needs to be re-written to use those `pub()`, `sub()` method in that case. I will work on that in some following weeks. (cc. @taoeffect)

Reply:
@SebinSong Sorry I meant maybe checking for `WebSocket.CLOSING`, in case the connection would drop just after login.

MDN says:
>If you call send() when the connection is in the CLOSING or CLOSED states, the browser will silently discard the data.

If this ever happens we likely want to re-send the message later